sql: descriptor validation overhaul

postamar opened the pull request:
sql: descriptor validation overhaul
    
    Previously, descriptor validation suffered from a few shortcomings:
    1. The methods to call were inconsistent across descriptor types, and
       the error messages they return are inconsistent as well.
    2. Those methods made ad-hoc use of catalog.DescGetter to walk through
       the descriptor reference graph, and could sometimes walk surprisingly
       far. This is particularly true for type descriptors, which hold
       references to other types.
    3. To complicate things further, DescGetter implementations which read
       descriptors from storage themselves perform validation as well.
       Although it is perfectly sensible to systematically validate
       descriptors when read, there is a circularity problem at hand which
       would benefit from being solved without melting the maintainer's
       brain in the process.
    4. The validation methods return an error type, making it awkward to
       return multiple errors. Tools like doctor would be made more useful
       if they could report more than only the first encountered error.

    Recently we introduced a change that added descriptor validation on
    write. This change involves validating all uncommitted descriptors in
    bulk and this in turn favours a bulk approach to reading all their
    referenced descriptors from storage.
    
    With this in mind, this commit adds a GetReferencedDescIDs method to
    catalog.Descriptor which returns the IDs of all descriptors referenced
    by the receiver. By calling this method on all descriptors we intend to
    validate and by reading all referenced descriptors in one kv.Batch and
    stuffing them in a catalog.MapDescGetter, we can now validate using this
    in-memory DescGetter instead.
    
    Turning validation into this multi-step process means it can no longer
    be invoked as a method call on the target descriptor. This commit moves
    the entry point to descriptor validation to the catalog.Validate
    function. The descriptor objects themselves still define the validation
    checks themselves but these are effectively made inaccessible from
    outside the catalog package. The rationale behind this is to enforce
    order in the sequence of checks performed and to enforce some uniformity
    in the formatting of the error messages. The checks are grouped into
    three tiers: internal consistency checks, cross-reference consistency
    checks, and transactional consistency checks.
    
    All this also helps reduce the number of call sites of descriptor
    validations as well as increase the scope of the validations. This has
    uncovered a few bugs related to schemas and temporary tables. This
    effort has also helped identify validation checks which have since been
    made redundant, either by other existing rules or by the new
    validation-on-write behaviour. Finally, this has allowed us to strip
    doctor of its own duplicated (and dodgy) validation logic and simply
    report all validation errors instead.
    
    Release justification: bug fixes and low-risk updates to new
    functionality (descriptor validation on write).
    
    Release note (cli change): The doctor tool can now report multiple
    descriptor validation failures per descriptor.

----

*Follow-up work*

- `WriteDescriptors` in `pkg/ccl/backupccl/restore_job.go`: should it validate schemas and types also? It seems like it should. When I add those, tests fail. I didn't look into it further. Probably out of scope for this change.
- ~Add linter check to enforce the use of `catalog.Validate` and `catalog.ValidateSelf`? It seems all too easy to call the `ValidateSelf` method on a `catalog.Descriptor` without being aware that it returns an `[]error` instead of an `error`.~ (edit: I since changed the way validation errors were collected, making a linter check unnecessary)
- Remove privileges surgery in `ValidateSelf` methods in favour of `NewFilledInImmutable`-like constructors. I don't like that validation functionality changes the underlying objects. (edit: this is actually really hard, and dangerous! giving up on this for now)
- ~I feel like my validation error wrapping is an improvement to the incoherent messaging that prevailed before but I'm not sure what's best to put in the prefix message. I settled on `errors.Wrapf(err, "%s %q", desc.TypeName(), desc.GetName())` but it could be anything else really. Do we prefer reporting names or IDs? I also don't know whether error message changes are considered to be contract-breaking, worthy of release note mentions, interfering with stability period, etc.~(edit: done like Andrew suggested below)

*Review guide*

The meat and potatoes of the change is in `pkg/sql/catalog/*.go`, best start reviewing from there, moving on to the `pkg/sql/catalog/*desc` packages. Notice that I moved the validation logic in `tabledesc` out of `structured.go` into a new file, `validate.go`. The diff in `structured.go` was going to be messy anyways despite there not being too many significant underlying changes: it's just calling `Report` with the errors, moving stuff around, removing redundancies, breaking down massive functions into smaller chunks, etc. The code still basically does the same thing. I moved tests cases around also.

Commit:
sql: descriptor validation overhaul

Previously, descriptor validation suffered from a few shortcomings:
1. The methods to call were inconsistent across descriptor types, and
   the error messages they return are inconsistent as well.
2. Those methods made ad-hoc use of catalog.DescGetter to walk through
   the descriptor reference graph, and could sometimes walk surprisingly
   far. This is particularly true for type descriptors, which hold
   references to other types.
3. To complicate things further, DescGetter implementations which read
   descriptors from storage themselves perform validation as well.
   Although it is perfectly sensible to systematically validate
   descriptors when read, there is a circularity problem at hand which
   would benefit from being solved without melting the maintainer's
   brain in the process.
4. The validation methods return an error type, making it awkward to
   return multiple errors. Tools like doctor would be made more useful
   if they could report more than only the first encountered error.

Recently we introduced a change that added descriptor validation on
write. This change involves validating all uncommitted descriptors in
bulk and this in turn favours a bulk approach to reading all their
referenced descriptors from storage.

With this in mind, this commit adds a GetReferencedDescIDs method to
catalog.Descriptor which returns the IDs of all descriptors referenced
by the receiver. By calling this method on all descriptors we intend to
validate and by reading all referenced descriptors in one kv.Batch and
stuffing them in a catalog.MapDescGetter, we can now validate using this
in-memory DescGetter instead.

Turning validation into this multi-step process means it can no longer
be invoked as a method call on the target descriptor. This commit moves
the entry point to descriptor validation to the catalog.Validate
function. The descriptor objects themselves still define the validation
checks themselves but these are effectively made inaccessible from
outside the catalog package. The rationale behind this is to enforce
order in the sequence of checks performed and to enforce some uniformity
in the formatting of the error messages. The checks are grouped into
three tiers: internal consistency checks, cross-reference consistency
checks, and transactional consistency checks.

All this also helps reduce the number of call sites of descriptor
validations as well as increase the scope of the validations. This has
uncovered a few bugs related to schemas and temporary tables. This
effort has also helped identify validation checks which have since been
made redundant, either by other existing rules or by the new
validation-on-write behaviour. Finally, this has allowed us to strip
doctor of its own duplicated (and dodgy) validation logic and simply
report all validation errors instead.

Release justification: This commit is safe for this release because it
consists of bug fixes and changes involving descriptor validation. If
anything these changes will help uncover descriptor corruptions which
were previously unnoticed, both because we're extending the validation
suite and because the doctor tool produces richer output.

Release note (cli change): The doctor tool can now report multiple
descriptor validation failures per descriptor.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/60775)
<!-- Reviewable:end -->

ajwerner commented:
> Add linter check to enforce the use of catalog.Validate and catalog.ValidateSelf? It seems all too easy to call the ValidateSelf method on a catalog.Descriptor without being aware that it returns an []error instead of an error.

This is interesting. I wonder if we should have it return a single error but have the implementation join the errors using `errors.CombineErrors` 

> Remove privileges surgery in ValidateSelf methods in favour of NewFilledInImmutable-like constructors. I don't like that validation functionality changes the underlying objects.

:+1: 

>  I also don't know whether error message changes are considered to be contract-breaking, worthy of release note mentions, interfering with stability period, etc.

No, our contracts, when they exist, are pgcode oriented. 

> I feel like my validation error wrapping is an improvement to the incoherent messaging that prevailed before but I'm not sure what's best to put in the prefix message. I settled on errors.Wrapf(err, "%s %q", desc.TypeName(), desc.GetName()) but it could be anything else really. Do we prefer reporting names or IDs? I also don't know whether error message changes are considered to be contract-breaking, worthy of release note mentions, interfering with stability period, etc.

Ideally we'd have fully qualified names, but, of course, at this layer we can't. I think i'd prefer both a name and ID. How about `errors.Wrapf(err, "%s %q (%d)", desc.TypeName(), desc.GetName(), desc.GetID())`

postamar commented:
This is ready for another look.

postamar commented on pkg/base/test_server_args.go:
```diff
@@ -128,9 +128,6 @@ type TestServerArgs struct {
 	EnableDemoLoginEndpoint bool
-
-	// If set, testing specific descriptor validation will be disabled. even if the server
-	DisableTestingDescriptorValidation bool
```
Note: I'm effectively reverting https://github.com/cockroachdb/cockroach/commit/dd9043993e47a385556da0ec4450740128f0810d, that framework is not actually used in any way.

postamar commented on pkg/sql/alter_database.go:
```diff
@@ -222,8 +221,7 @@ func (n *alterDatabaseAddRegionNode) startExec(params runParams) error {
 	// we're using an internal call to addEnumValue above which doesn't perform validation.
-	dg := catalogkv.NewOneLevelUncachedDescGetter(params.p.txn, params.ExecCfg().Codec)
-	if err := typeDesc.Validate(params.ctx, dg); err != nil {
+	if err := validateDescriptor(params.ctx, params.p, typeDesc); err != nil {
```
Note: This convenience function is defined in planner.go: https://github.com/cockroachdb/cockroach/pull/60775/files#diff-d4a3b2d400fd963e130106b8942783b2abbf757310a2f3904ace7cc5dc4052bfR794f

postamar commented on pkg/sql/alter_database.go:
```diff
@@ -536,12 +534,6 @@ func (n *alterDatabasePrimaryRegionNode) switchPrimaryRegion(params runParams) e
-	if err := typeDesc.Validate(params.ctx, dg); err != nil {
-		return err
-	}
-
```
Note: I believe this was made redundant by validation-on-write.

postamar commented on pkg/sql/alter_table.go:
```diff
@@ -281,7 +280,7 @@ func (n *alterTableNode) startExec(params runParams) error {
 				var err error
 				params.p.runWithOptions(resolveFlags{contextDatabaseID: n.tableDesc.ParentID}, func() {
-					info, infoErr := n.tableDesc.GetConstraintInfo(params.ctx, nil)
+					info, infoErr := n.tableDesc.GetConstraintInfo()
```
Note: I changed the signatures of GetConstraintInfo and associated methods to cut down on passing nil DescGetters. https://github.com/cockroachdb/cockroach/pull/60775/files#diff-e971105bc05467ef391df9792202d3a8b712d269b5f193a0d94b700414ef281eL180

postamar commented on pkg/sql/alter_table_locality.go:
```diff
@@ -135,7 +134,7 @@ func (n *alterTableSetLocalityNode) alterTableLocalityGlobalToRegionalByTable(
 	// Finalize the alter by writing a new table descriptor and updating the zone
 	// configuration.
-	if err := n.validateAndWriteNewTableLocalityAndZoneConfig(
+	if err := n.writeNewTableLocalityAndZoneConfig(
```
Note: validation made implicit by validate-on-write.

postamar commented on pkg/sql/alter_type.go:
```diff
@@ -126,12 +126,6 @@ func (n *alterTypeNode) startExec(params runParams) error {
-	if err := n.desc.Validate(params.ctx, dg); err != nil {
-		return err
-	}
-
```
Note: validation made implicit by validate-on-write.

postamar commented on pkg/sql/catalog/dbdesc/database_test.go:
```diff
@@ -79,3 +80,72 @@ func TestMakeDatabaseDesc(t *testing.T) {
 	}
 }
+
+func TestValidateDatabaseDesc(t *testing.T) {
```
Note: moved here from `tabledesc`.

postamar commented on pkg/sql/catalog/descpb/privilege.go:
```diff
@@ -218,7 +218,11 @@ func (p *PrivilegeDescriptor) Revoke(
+	if *ptr == nil {
+		*ptr = &PrivilegeDescriptor{}
+	}
+	p := *ptr
```
Note: this was done mainly to accomodate tests. I think it's also nicer though.

postamar commented on pkg/ccl/streamingccl/streamclient/random_stream_client.go:
```diff
@@ -223,7 +224,7 @@ func (m *randomStreamClient) getDescriptorAndNamespaceKVForTableID(
 		tableID,
 		fmt.Sprintf(RandomStreamSchemaPlaceholder, tableName),
-		&descpb.PrivilegeDescriptor{},
+		descpb.NewDefaultPrivilegeDescriptor(security.RootUserName()),
```
Note: I'll revert this, I committed this by mistake.

postamar commented on pkg/sql/catalog/descriptor.go:
```diff
@@ -73,14 +73,18 @@ type Descriptor interface {
 
-	// Validate is like ValidateSelf but with additional cross-reference checks.
-	Validate(ctx context.Context, descGetter DescGetter) error
+	// ValidateSelfAndCrossReferences performs cross-reference checks.
```
Note: I'll fix this typo.

postamar commented on pkg/sql/catalog/descs/collection.go:
```diff
@@ -1072,6 +1082,16 @@ func (tc *Collection) GetMutableDescriptorByIDWithFlags(
+	log.VEventf(ctx, 2, "planner getting immutable descriptor for id %d", id)
+	return tc.getDescriptorByID(ctx, txn, id, flags, false /* mutable */)
+}
+
```
Note: this is used by the repair functions.

postamar commented on pkg/sql/catalog/tabledesc/safe_format_test.go:
```diff
@@ -38,11 +39,11 @@ func TestSafeMessage(t *testing.T) {
 	}{
 		{
-			id:       12,
+			id:       112,
```
Note: Had to bump this ID as it was in the system table range.

postamar commented on pkg/sql/create_view.go:
```diff
@@ -92,25 +91,29 @@ func (n *createViewNode) startExec(params runParams) error {
-			)
-			persistence = tree.PersistenceTemporary
+		if !n.persistence.IsTemporary() && backRefMutable.Temporary {
+			hasTempBackref = true
```
Note: it seemed to me that the notice could end up being printed multiple times so I rewrote this.

postamar commented on pkg/sql/create_view.go:
```diff
@@ -92,25 +91,29 @@ func (n *createViewNode) startExec(params runParams) error {
 		backRefMutables[id] = backRefMutable
 	}
+	if hasTempBackref {
+		n.persistence = tree.PersistenceTemporary
```
Note: this was missing and was a bug, we need to update the persistence. Is it bad that I mutate `n.persistence` instead of `persistence`?

postamar replied:
Also, I added a test in `tabledesc` to cover this, search for `// Temporary tables.`.

postamar replied:
The `temp_table` logic test fails here: https://github.com/postamar/cockroach/blob/master/pkg/sql/logictest/testdata/logic_test/temp_table#L143
with
```
CREATE VIEW upgrade_temp_view AS SELECT a FROM temp_table;
	-- FAIL
E210224 16:02:27.451156 3907 sql/logictest/logic.go:3563  [-] 2  
E210224 16:02:27.451156 3907 sql/logictest/logic.go:3563  [-] 2 +testdata/logic_test/temp_table:143: CREATE VIEW upgrade_temp_view AS SELECT a FROM temp_table
E210224 16:02:27.451156 3907 sql/logictest/logic.go:3563  [-] 2 +expected success, but found
E210224 16:02:27.451156 3907 sql/logictest/logic.go:3563  [-] 2 +(XXUUU) relation "upgrade_temp_view" (71): referenced schema descriptor 65 not found
E210224 16:02:27.451156 3907 sql/logictest/logic.go:3563  [-] 2 +validate.go:221: in GetSchemaDescriptor()
```
This happens because the view is on a temp table and therefore should also have its persistence set to `tree.PersistenceTemporary` or else during validation it will try to cross-reference a non-existent schema. Previously the persistence was _not_ set because further down in this file we create the table descriptor using `n.persistence` instead of the local `persistence` variable.

postamar commented on pkg/sql/doctor/doctor.go:
```diff
@@ -339,18 +287,13 @@ func ExamineJobs(
+	// The prefix has the same format as the validation error wrapper.
+	msgPrefix := fmt.Sprintf("%s %q (%d): ", desc.TypeName(), desc.GetName(), desc.GetID())
+	if msg[:len(msgPrefix)] == msgPrefix {
+		msgPrefix = ""
```
Note: this test is required if we want to make the namespace table report messages share the same format as the descriptor validation error messages.

postamar commented on pkg/sql/doctor/doctor_test.go:
```diff
@@ -60,8 +60,28 @@ var validTableDesc = &descpb.Descriptor{
+			descpb.MaybeFixPrivileges(typ.GetID(), &typ.Privileges)
+		}
+	}
+	res, err := protoutil.Marshal(desc)
```
Note: I really couldn't be bothered to update all of the descriptor declarations so I did this instead.

postamar commented on pkg/sql/repair.go:
```diff
@@ -380,32 +385,24 @@ func (p *planner) UnsafeUpsertNamespaceEntry(
+		case catalog.SchemaDescriptor:
+			invalid = parentID == descpb.InvalidID || parentSchemaID != descpb.InvalidID
+		case catalog.DatabaseDescriptor:
+			invalid = parentID != descpb.InvalidID || parentSchemaID != descpb.InvalidID
```
Note: this is a refactor.

postamar commented on pkg/sql/schema_changer.go:
```diff
@@ -1268,15 +1268,6 @@ func (sc *SchemaChanger) done(ctx context.Context) error {
-						ctx,
-						dg,
-					); err != nil {
-						return err
```
Note: made redundant by validate-on-write.

postamar commented on pkg/sql/catalog/tabledesc/structured.go:
```diff
@@ -75,16 +71,18 @@ var ErrIndexGCMutationsList = errors.New("index in GC mutations list")
 func NewCreatedMutable(tbl descpb.TableDescriptor) *Mutable {
-	return &Mutable{wrapper: wrapper{TableDescriptor: tbl}}
+	m, _ := NewFilledInExistingMutable(context.TODO(), nil /* DescGetter */, false /* skipFKsWithMissingTable */, &tbl)
+	return &Mutable{wrapper: m.wrapper}
```
Note: this is mainly to make sure that the privileges are set from the get-go.

postamar commented on pkg/sql/catalog/tabledesc/structured.go:
```diff
@@ -1659,26 +1280,19 @@ func FormatTableLocalityConfig(c *descpb.TableDescriptor_LocalityConfig, f *tree
 }
 
-// ValidateTableLocalityConfig validates whether the descriptor's locality
+// validateTableLocalityConfig validates whether the descriptor's locality
```
Note: I forgot to move this to `validate.go`, will do.

postamar commented on pkg/sql/catalog/tabledesc/structured.go:
```diff
@@ -1252,387 +1254,6 @@ func (desc *Mutable) OriginalVersion() descpb.DescriptorVersion {
-// Validate performs ValidateSelf and then validates that
-// each reference to another table is resolvable and that the necessary back
-// references exist.
-func (desc *wrapper) Validate(ctx context.Context, descGetter catalog.DescGetter) error {
```
Note: I moved this and the other validation functions to a new source file `validate.go`. The diffs were pretty gnarly anyway and as it turns out this also helps slim down `structured.go` which is far too big a source file.

postamar commented on pkg/sql/catalog/tabledesc/structured_test.go:
```diff
@@ -137,1459 +133,6 @@ func TestAllocateIDs(t *testing.T) {
 		t.Fatalf("expected %s, but found %s", a, b)
 	}
 }
-func TestValidateDatabaseDesc(t *testing.T) {
```
Note: moved this to `dbdesc`.

postamar commented on pkg/sql/catalog/tabledesc/validate.go:
```diff
@@ -0,0 +1,1207 @@
+// Copyright 2021 The Cockroach Authors.
```
Note: here lives the table validation logic now. Unlike for type descriptors the table validation logic itself didn't need a lot of changes so I sincerely don't think this should impede review.

ajwerner reviewed (commented):
This is really nice. I like the error reporter idea; it's much more flexible.

postamar commented on pkg/sql/catalog/dbdesc/database_desc.go:
```diff
@@ -271,55 +268,76 @@ func (desc *Immutable) ForEachSchemaInfo(
+	// Fill in any incorrect privileges that may have been missed due to mixed-versions.
+	// TODO(mberhault): remove this in 2.1 (maybe 2.2) when privilege-fixing migrations have been
+	// run again and mixed-version clusters always write "good" descriptors.
+	descpb.MaybeFixPrivileges(desc.ID, &desc.Privileges)
```
Quite right! Filed https://github.com/cockroachdb/cockroach/issues/61070.

postamar commented on pkg/sql/catalog/dbdesc/database_desc.go:
```diff
@@ -271,55 +268,76 @@ func (desc *Immutable) ForEachSchemaInfo(
+func (desc *Immutable) ValidateCrossReferences(
+	_ catalog.ValidationErrorAccumulator, _ catalog.ValidationDescGetter,
+) {
+	// No-op.
```
Right! Dunno why it didn't occur to me. I'll do it.

postamar commented on pkg/sql/catalog/desc_getter.go:
```diff
@@ -25,40 +25,33 @@ type DescGetter interface {
-	typ, ok := desc.(TypeDescriptor)
-	if !ok {
-		return nil, ErrDescriptorNotFound
+// AsBatchDescGetter is a convenience function for turning a DescGetter into a
```
You're right, it doesn't help anywhere. I think I just forgot to clean this up. Will fix.

postamar commented:
Thanks a lot for the speedy review @ajwerner. I believe I've addressed your concerns in the latest revision I pushed. I also added some unit tests in `dbdesc` and `schemadesc` which were missing. 

Random thought: test coverage is always going to be spotty unless we enforce it somehow. Ideally I'd like the test suite to fail should there exist a specific validation error which doesn't have a corresponding unit test. Perhaps we could use the type system to this effect:
- force `Report` to only accept a specific error subtype for validations, 
- which each need to be registered, 
- write a test which matches the contents of the registry with the unit test cases.
- those test cases contain an expected error field, right now it's a string but it could be of a type that matches the registry entry somehow.

This validation error subtyping might have other interesting benefits like making the error message formatting more consistent, allowing for `errors.Is` assertions, etc.

ajwerner reviewed (commented):
> This validation error subtyping might have other interesting benefits like making the error message formatting more consistent, allowing for errors.Is assertions, etc.

This is a fun idea. I think the way I'd do this is to have a hook in a testing knob to wrap the error reporter and then accumulate all of the errors generated through the test and then at the end of the test cross check that against a list of errors we expect to exist. We could do this leveraging types, or not. We could enforce this all further by saying that all of the error types need to be registered in some form or another at init time. 

Didn't finish but here's a quick comment

> Reviewed 26 of 69 files at r1, 1 of 5 files at r2, 14 of 23 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60775#-:-MUKMOeH4vOXVIlQMFcZ:b-v30gg3)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @miretskiy, and @postamar)

___
*[pkg/sql/scrub.go, line 433 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/60775#-MUKfW75ClOiBYziqyRc:-MUKfW75ClOiBYziqyRd:bksiazr) ([raw file](https://github.com/cockroachdb/cockroach/blob/d443b8212d98079cf2f6f6614893dd310781f66b/pkg/sql/scrub.go#L433)):*
> ```Go
> 		table, ok := desc.(catalog.TableDescriptor)
> 		if !ok {
> 			return nil, catalog.ErrDescriptorNotFound
> ```

any interest in telling us which descriptor wasn't found with an `errors.Wrapf`?

Scrub is sort of deprecated so not sure it matters too much. 
___
*[pkg/sql/catalog/validate.go, line 139 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/60775#-MUKgKlHB1hUTIFMRpVC:-MUKgKlIDpBUSxfWoJCu:b-ki061e) ([raw file](https://github.com/cockroachdb/cockroach/blob/d443b8212d98079cf2f6f6614893dd310781f66b/pkg/sql/catalog/validate.go#L139)):*
<details><summary><i alt="summary">Quoted 4 lines of code…</i></summary>

> ```
> type validationErrorAccumulatorImpl struct {
> 	wrapPrefix string
> 	errors     []error
> }
> ```
</details>

minor nit, make a separate type for the `ValidationErrors` implementation.

```
type validationErrors struct {
    errors []error
}

func (*validationErrors) sealed() {}

type validationErrorAccumulator struct {
    validationErrors
    wrapPrefix string
}
```

Also, I feel like the pattern where an unexported struct which implements an exported interface of the same name is pretty clear that it's the impl. 


<!-- Sent from Reviewable.io -->

postamar reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60775)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @miretskiy, and @postamar)

___
*[pkg/sql/scrub.go, line 433 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/60775#-MUKfW75ClOiBYziqyRc:-MUL_Iqo2VcbYGHkKy4Q:b6j92lo) ([raw file](https://github.com/cockroachdb/cockroach/blob/d443b8212d98079cf2f6f6614893dd310781f66b/pkg/sql/scrub.go#L433)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

any interest in telling us which descriptor wasn't found with an `errors.Wrapf`?

Scrub is sort of deprecated so not sure it matters too much. 
</blockquote></details>

Sure! In fact I went a bit further and made use of ErrDescNotFound in ValidationDescGetter and friends.
___
*[pkg/sql/catalog/validate.go, line 139 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/60775#-MUKgKlHB1hUTIFMRpVC:-MUL_T9U0BxrPBMqvnsv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d443b8212d98079cf2f6f6614893dd310781f66b/pkg/sql/catalog/validate.go#L139)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> type validationErrorAccumulatorImpl struct {
> 	wrapPrefix string
> 	errors     []error
> }
> ```
minor nit, make a separate type for the `ValidationErrors` implementation.

```
type validationErrors struct {
    errors []error
}

func (*validationErrors) sealed() {}

type validationErrorAccumulator struct {
    validationErrors
    wrapPrefix string
}
```

Also, I feel like the pattern where an unexported struct which implements an exported interface of the same name is pretty clear that it's the impl. 
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

Commit:
doctor: add validation runtime error recovery

Previously, doctor could crash when validating corrupted descriptors.
The validation code is usually cautious enough to not dereference nils
or the like but this isn't enforced in any way.

This commit protects the validation calls in doctor with panic recovery,
allowing doctor to continue and validate the remaining descriptors.

Release justification: non-production code change.
Release note: None

postamar commented:
I rebased w/ linter error fix + a couple more validation test cases. I also added a new small commit which protects doctor validations from runtime errors like nil dereferences. I'm uncomfortable about the possibility of a customer running doctor to debug a cluster and doctor itself crashing 😬

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> just a couple of questions

> Reviewed 14 of 69 files at r1, 2 of 5 files at r2, 4 of 23 files at r3, 23 of 23 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60775#-:-MUVlZodDDe4Xi3WQprt:bjvgq1s)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @miretskiy, and @postamar)

___
*[pkg/sql/catalog/dbdesc/database_desc.go, line 343 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/60775#-MUVj5hv33Bd7dsqNj9a:-MUVj5hv33Bd7dsqNj9b:b394k5n) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad74fef98fbb126b6ecc12993934518f3acac031/pkg/sql/catalog/dbdesc/database_desc.go#L343)):*
> ```Go
> 		report := func(err error) {
> 			vea.Report(errors.Wrapf(err, "schema mapping entry %q (%d)",
> 				errors.Safe(schemaName), errors.Safe(schemaInfo.ID)))
> ```

descpb.ID is always safe to format; no need to wrap values of that type in Safe.

https://github.com/cockroachdb/cockroach/blob/de72eae2c0a984d00d404f22850b2642c29a877e/pkg/sql/catalog/descpb/structured.go#L41
___
*[pkg/sql/catalog/descs/collection.go, line 1453 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/60775#-MUVjd9y0yxP7Xr6YR_X:-MUVjd9z8zcZYnpvC0TA:b-tdwmfe) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad74fef98fbb126b6ecc12993934518f3acac031/pkg/sql/catalog/descs/collection.go#L1453)):*
> ```Go
> }
> 
> // ValidateUncommittedDescriptors validates all uncommitted descriptors
> ```

maybe comment that this goes to the store for all of the descriptors?

When I was initially thinking about this stuff, I was thinking that we wouldn't go re-read them all but in practice this seems all fine and good given the new parallelism. 
___
*[pkg/sql/catalog/tabledesc/table.go, line 179 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/60775#-MUVkS1r96pTNKO2kQis:-MUVkS1r96pTNKO2kQit:b-fd30gl) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad74fef98fbb126b6ecc12993934518f3acac031/pkg/sql/catalog/tabledesc/table.go#L179)):*
> ```Go
> }
> 
> // GetConstraintInfo returns a summary of all constraints on the table.
> ```

does it matter that now the constraint details won't have descriptors? Was anybody using that? If not, can we deprecate that field altogether and then remove it in the subsequent release? 
___
*[pkg/sql/pgwire/testdata/pgtest/notice, line 58 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/60775#-MUVlRGJD9VjAz_QRhxA:-MUVlRGJD9VjAz_QRhxB:bx88ec6) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad74fef98fbb126b6ecc12993934518f3acac031/pkg/sql/pgwire/testdata/pgtest/notice#L58)):*
> ```
> CommandComplete
> ----
> {"Severity":"NOTICE","SeverityUnlocalized":"","Code":"00000","Message":"the data for dropped indexes is reclaimed asynchronously","Detail":"","Hint":"The reclamation delay can be customized in the zone configuration for the table.","Position":0,"InternalPosition":0,"InternalQuery":"","Where":"","SchemaName":"","TableName":"","ColumnName":"","DataTypeName":"","ConstraintName":"","File":"drop_index.go","Line":519,"Routine":"dropIndexByName","UnknownFields":null}
> ```

:( every time this happens I cry a little inside 


<!-- Sent from Reviewable.io -->

postamar reviewed (commented):
Thanks for looking at this.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60775#-:-MUi73p6DQ8fh-OBGMxR:baqh0x5)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner, @miretskiy, and @postamar)

___
*[pkg/sql/catalog/dbdesc/database_desc.go, line 343 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/60775#-MUVj5hv33Bd7dsqNj9a:-MUi4nCD34Nz5BiqBdqb:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad74fef98fbb126b6ecc12993934518f3acac031/pkg/sql/catalog/dbdesc/database_desc.go#L343)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

descpb.ID is always safe to format; no need to wrap values of that type in Safe.

https://github.com/cockroachdb/cockroach/blob/de72eae2c0a984d00d404f22850b2642c29a877e/pkg/sql/catalog/descpb/structured.go#L41
</blockquote></details>

Done.
___
*[pkg/sql/catalog/descs/collection.go, line 1453 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/60775#-MUVjd9y0yxP7Xr6YR_X:-MUi4q2yEaBIeBElUKhn:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad74fef98fbb126b6ecc12993934518f3acac031/pkg/sql/catalog/descs/collection.go#L1453)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

maybe comment that this goes to the store for all of the descriptors?

When I was initially thinking about this stuff, I was thinking that we wouldn't go re-read them all but in practice this seems all fine and good given the new parallelism. 
</blockquote></details>

Done.
___
*[pkg/sql/catalog/tabledesc/table.go, line 179 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/60775#-MUVkS1r96pTNKO2kQis:-MUi5_5b1TljV1SV4YTL:b-22kb11) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad74fef98fbb126b6ecc12993934518f3acac031/pkg/sql/catalog/tabledesc/table.go#L179)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

does it matter that now the constraint details won't have descriptors? Was anybody using that? If not, can we deprecate that field altogether and then remove it in the subsequent release? 
</blockquote></details>

Those do appear to be used in `information_schema.go`, `pg_catalog.go` and such places.
___
*[pkg/sql/pgwire/testdata/pgtest/notice, line 58 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/60775#-MUVlRGJD9VjAz_QRhxA:-MUi6cy7-ZnrdKMDkWtT:b-443bwe) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad74fef98fbb126b6ecc12993934518f3acac031/pkg/sql/pgwire/testdata/pgtest/notice#L58)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

:( every time this happens I cry a little inside 
</blockquote></details>

If I had a dollar every time this happened I'd be able to buy myself a latte every month or two!


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 24 of 24 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60775#-:-MUiF50Y-Y8f4Vtthcqt:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @miretskiy, and @postamar)



<!-- Sent from Reviewable.io -->

postamar commented:
bors r+
